[user_ldap] Update profile from LDAP fields

march42 opened the pull request:
* Resolves: #32085

## Summary
Implement feature, to update user profile data from LDAP attributes.
Add configuration items for LDAP attribute mapping (Administration - LDAP/AD integration - Advanced - User Profile Attributes)
Add methods for user profile properties to lib/private/User/User.php
Add connections for LDAP attributes to user profile properties

## TODO
- [ ] Testing with various setting and scenario

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [x] [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
![Screenshot (Nextcloud 2)](https://user-images.githubusercontent.com/5712929/217013965-ec751920-6d14-45de-9455-5044e1b63eea.png)
![Screenshot (1)](https://user-images.githubusercontent.com/5712929/217010632-5ab0f259-1d96-4f7f-8220-b455f9386cb3.png)

- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
EDIT: [documentation PR 9614](https://github.com/nextcloud/documentation/pull/9614)
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

march42 committed:
feature addition: [user_ldap] update user profile from LDAP; WIP work-in-progress; TODO update profile

Signed-off-by: Marc Hefter <marchefter@march42.net>

march42 committed:
feature addition: [user_ldap] update user profile from LDAP; WIP; fixing some uggly copy-and-paste errors; testing functionality; preparing and editing the documentation

Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
feature addition: [user_ldap] update user profile from LDAP

Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
added user profile scope setting

Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
bugfix: changed forgotten userName, to getUsername

Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
fixed error: Array to string conversion at user_ldap/lib/User/User.php#639

Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
cleanup: removed unnecessary imports of IAccountManager, AccountManager, IDBConnection

Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
 
+		//User profile visibility
+		$profileScope = $this->connection->ldapProfileScope;
+		if (is_null($profileScope) || '' === $profileScope || 'unset' === $profileScope) {
```
```suggestion
		if ($profileScope === '' || $profileScope === 'unset') {
```
But when is it 'unset' ?

march42 replied:
When choosen in the advanced configuration tab. The administrator chooses profileScope from unset, v2-private, v2-local, v2-federated, v2-published. If using 'unset' it means the user can set these self in the profile.

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
+		if (isset($ldapEntry[$attr])) {
+			$this->updateProfile(self::USER_PREFKEY_PHONE, $ldapEntry[$attr][0], $profileScope);
+		}
+		unset($attr);
```
```suggestion
```
What are these unset for?

march42 replied:
habit: cleaning up the memory used, to ease garbage collection
I would remove all but the last, to unclutter it

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +580,69 @@ private function verifyQuotaValue(string $quotaValue) {
+	/**
+	 * fetches values from LDAP and stores it as Nextcloud user value
+	 * @param string $valueFromLDAP if known, to save an LDAP read request
+	 * @return null
+	 */
+	private function updateProfile(string $property, $valueFromLDAP, $scope=null) {
```
```suggestion
	 */
	private function updateProfile(string $property, $valueFromLDAP, ?string $scope=null): void {
```

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +580,69 @@ private function verifyQuotaValue(string $quotaValue) {
+			}
+			$this->logger->debug('property updated: '.$profileProperty.'='.$value.' for user '.$this->getUsername().'', ['app' => 'user_ldap']);
+			$this->config->setUserValue($this->getUsername(), 'user_ldap', $property, $value);
+			return $value;
```
```suggestion
			return;
```

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +580,69 @@ private function verifyQuotaValue(string $quotaValue) {
+		} else {
+			// FIXME: I decided, to leave profile untouched, if attribute gets removed from LDAP
+			$this->config->deleteUserValue($this->getUsername(), 'user_ldap', $property);
+			return '';
```
```suggestion
			return;
```

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +580,69 @@ private function verifyQuotaValue(string $quotaValue) {
+		if ($propertyValue && isset($propertyValue[0])) {
+			$value = $propertyValue[0];
+			try {
+				$user = $this->userManager->get($this->uid);
```
For each configured profile property you will fetch the user again. You should prepare an array of profile values to apply and give that to your method, which could then fetch the user only once.

march42 replied:
Sounds good, I will look into this.

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +580,69 @@ private function verifyQuotaValue(string $quotaValue) {
+			return $value;
+		} else {
+			// FIXME: I decided, to leave profile untouched, if attribute gets removed from LDAP
+			$this->config->deleteUserValue($this->getUsername(), 'user_ldap', $property);
```
profile properties need to be set both on user and in config?

march42 replied:
No, I remove this. I used it to easily check the data with `occ user:setting` command.

come-nc commented on lib/private/User/User.php:
```diff
@@ -587,4 +588,104 @@ public function triggerChange($feature, $value = null, $oldValue = null) {
+	 * @return string|null AccountProperty value
+	 * @throws InvalidArgumentException when the property name is invalid or null
+	 */
+	public function getProfilePropertyValue($property): ?string {
```
Are you sure all those proxy methods are needed? What prevents user_ldap from getting the account itself?

come-nc replied:
You should use `IAccountManager::getAccount($user)` to get the account,
then `IAccount::setProperty` for each property, and then `IAccountManager::updateAccount($account)`.

There is no need to add methods to OCP\IUser for this.

come-nc reviewed (commented):
The feature will be popular I think but the code needs to be refined.
Also we must be very careful about performance in user_ldap, this should have no performance impact when not used.

@Pytal Can you review this? It’s a lot related to the profile system.

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
+		if (isset($ldapEntry[$attr])) {
+			$this->updateProfile(self::USER_PREFKEY_WEBSITE, $ldapEntry[$attr][0], $profileScope);
+		}
+		unset($attr);
```
```suggestion
```

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
+		if (isset($ldapEntry[$attr])) {
+			$this->updateProfile(self::USER_PREFKEY_ADDRESS, $ldapEntry[$attr][0], $profileScope);
+		}
+		unset($attr);
```
```suggestion
```

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
+		if (isset($ldapEntry[$attr])) {
+			$this->updateProfile(self::USER_PREFKEY_TWITTER, $ldapEntry[$attr][0], $profileScope);
+		}
+		unset($attr);
```
```suggestion
```

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
+		if (isset($ldapEntry[$attr])) {
+			$this->updateProfile(self::USER_PREFKEY_ORGANISATION, $ldapEntry[$attr][0], $profileScope);
+		}
+		unset($attr);
```
```suggestion
```

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
+		if (isset($ldapEntry[$attr])) {
+			$this->updateProfile(self::USER_PREFKEY_ROLE, $ldapEntry[$attr][0], $profileScope);
+		}
+		unset($attr);
```
```suggestion
```

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +245,60 @@ public function processAttributes($ldapEntry) {
+		if (isset($ldapEntry[$attr])) {
+			$this->updateProfile(self::USER_PREFKEY_HEADLINE, $ldapEntry[$attr][0], $profileScope);
+		}
+		unset($attr);
```
```suggestion
```

march42 reviewed (commented):
cleanup: remove unneeded unset

Commit:
CleanUp: remove unneeded UNSET, to unclutter code

Signed-off-by: march42 <marchefter@gmail.com>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
Fix: Parameter $scope has no provided type

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

Pytal commented on lib/public/IUser.php:
```diff
@@ -270,4 +271,47 @@ public function getQuota();
+	 * @param string $property name see IAccountManager::PROPERTY_*
+	 * @return string AccountProperty value
+	 * @throws InvalidArgumentException when the property name is invalid or null
+	 * @since 25.0.0
```
```suggestion
	 * @since 27.0.0
```

Pytal commented on lib/public/IUser.php:
```diff
@@ -270,4 +271,47 @@ public function getQuota();
+	 * @param string $property name see IAccountManager::PROPERTY_*
+	 * @return string AccountProperty scope IAccountManager::SCOPE_*
+	 * @throws InvalidArgumentException when the property name is invalid or null
+	 * @since 25.0.0
```
```suggestion
	 * @since 27.0.0
```

Pytal commented on lib/public/IUser.php:
```diff
@@ -270,4 +271,47 @@ public function getQuota();
+	 * @param string $property name see IAccountManager::PROPERTY_*
+	 * @return string AccountProperty verification status IAccountManager::NOT_VERIFIED/VERIFICATION_IN_PROGRESS/VERIFIED
+	 * @throws InvalidArgumentException when the property name is invalid or null
+	 * @since 25.0.0
```
```suggestion
	 * @since 27.0.0
```

Pytal commented on lib/public/IUser.php:
```diff
@@ -270,4 +271,47 @@ public function getQuota();
+	 * @param string $verified AccountProperty verified
+	 * @return void
+	 * @throws InvalidArgumentException when the property name is invalid or null
+	 * @since 25.0.0
```
```suggestion
	 * @since 27.0.0
```

Pytal commented on lib/private/User/User.php:
```diff
@@ -587,4 +588,104 @@ public function triggerChange($feature, $value = null, $oldValue = null) {
+
+	/**
+	 * @param string $property name of the AccountProperty
+	 * @return string|null AccountProperty value
```
```suggestion
	 * @return ?string AccountProperty value
```

Pytal commented on lib/private/User/User.php:
```diff
@@ -587,4 +588,104 @@ public function triggerChange($feature, $value = null, $oldValue = null) {
+
+	/**
+	 * @param string $property name of the AccountProperty
+	 * @return string|null AccountProperty scope
```
```suggestion
	 * @return ?string AccountProperty scope
```

Pytal commented on lib/private/User/User.php:
```diff
@@ -587,4 +588,104 @@ public function triggerChange($feature, $value = null, $oldValue = null) {
+
+	/**
+	 * @param string $property name of the AccountProperty
+	 * @return string|null AccountProperty verified
```
```suggestion
	 * @return ?string AccountProperty verified
```

Commit:
CleanUp: tidy checking for empty profileScope

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
CleanUp: correct version to 27 and return value specification

Co-authored-by: Pytal <24800714+Pytal@users.noreply.github.com>
Signed-off-by: Marc Hefter <marchefter@march42.net>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

Commit:
CleanUp: removed redundant condition if block

Signed-off-by: Marc Hefter <marchefter@gmail.com>

Commit:
CleanUp: removed redundant storing profile attributes in user settings

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
handling updateProfile with array of values

using an array to buffer profile updates, like suggested by @come-nc
clean some code and remove unneccessary redundancy
added the Fediverse profile property

Co-Authored-By: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
fixing Psalm messages

Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 commented:
I reworked the code:

- prepare and propagate an array with the profile data from LDAP and fetching user just once. Like @come-nc suggested.
- added Fediverse profile property
- did some cleanup and simplification
- fixed messages from psalm

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +233,70 @@ public function processAttributes($ldapEntry) {
+			$profileValues[\OCP\Accounts\IAccountManager::PROPERTY_FEDIVERSE] = $ldapEntry[$attr][0];
+		}
+		// Update user profile
+		if(0 < count($profileValues)) {
```
```suggestion
		if (count($profileValues) > 0) {
```

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -231,6 +233,70 @@ public function processAttributes($ldapEntry) {
+		// Update user profile
+		if(0 < count($profileValues)) {
+			$this->updateProfile($profileValues, $profileScope);
+			unset($profileValues);
+		}
+		unset($profileScope);
+		unset($attr);
```
I’m still not comfortable with those unset, we are not micromanaging memory anywhere else in our code like this.

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +578,52 @@ private function verifyQuotaValue(string $quotaValue) {
+			return;
+		}
+		// check if parameter array is empty
+		if(0 == count($profileValues)) {
```
```suggestion
		if (count($profileValues) == 0) {
```

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +578,52 @@ private function verifyQuotaValue(string $quotaValue) {
+			// check and update profile properties
+			/** @var string $propertyValue */
+			$propertyValue = [$valueFromLDAP];
+			if (isset($propertyValue[0])) {
```
This is always true, no?

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +578,52 @@ private function verifyQuotaValue(string $quotaValue) {
+		foreach($profileValues as $property => $valueFromLDAP) {
+			$this->logger->debug('user profile data ('.$property.') from LDAP '.$this->dn, ['app' => 'user_ldap']);
+			// check and update profile properties
+			/** @var string $propertyValue */
```
```suggestion
			/** @var string[] $propertyValue */
```

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +578,52 @@ private function verifyQuotaValue(string $quotaValue) {
+			/** @var string $propertyValue */
+			$propertyValue = [$valueFromLDAP];
+			if (isset($propertyValue[0])) {
+				/** @var string $value */
```
```suggestion
```

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +578,52 @@ private function verifyQuotaValue(string $quotaValue) {
+							$this->logger->debug('property updated: '.$property.'='.$value.' for user '.$this->getUsername().'', ['app' => 'user_ldap']);
+						}
+				} catch (PropertyDoesNotExistException $e) {
+					$this->logger->error('property does not exist: '.$property.' for user '.$this->getUsername().'', ['app' => 'user_ldap']);
```
```suggestion
					$this->logger->error('property does not exist: '.$property.' for user '.$this->getUsername().'', ['app' => 'user_ldap', 'exception' => $e]);
```

come-nc reviewed (changes requested):
See my last comment, please go through the accountmanager instead of adding methods to `OCP\User`.

Also, could you please detail how this will behave for multivalue fields from LDAP?

Regarding scope, if I understand correctly the admin can force a given scope for all profile data synced from LDAP?
What happens if the user then changes the scope?

Is it possible for the admin to leave the scope option empty to let users decide the scope of each information?

come-nc commented:
> I have found, that some commits are not correctly signed-off. Seems to be like all commits through GitHub Desktop. So I would probably best rebase the branch before you finally merge it. Any suggestions?

You can use `rebase` with `--signoff` option.
For instance `git rebase HEAD~25 --signoff` will add a signoff to the last 25 commits.

come-nc commented:
> Now, you got me a little pissed, because you were unspecific and I think I'm not supposed to read your mind. I'll be specific, so we could find an agreement on how to handle. First of all. I feel unappreciated, even unseen. I created my branch in May 2022, after issue 32085 was opened. see [#32085 (comment)](https://github.com/nextcloud/server/issues/32085#issuecomment-1119313753) I invested time, to package my quick'n'dirty hack and pretty it up, before committing to GitHub. I updated the documentation, run numerous test and checked different scenarios. Nobody seemed to care, until February 2023, when I was asked to make a pull request. 

I’m sorry you feel that way, that was not my intention of course.
I do not follow feature requests in the issues closely as we already have a lot on our plate. So I only became aware of this when the PR was open and I got pinged for a review.
I’m used to reviewing code from collegues and being picky about it, to try to merge the best code as possible. Sorry if that felt like too much complains, your work is appreciated of course. As I previously stated I think this feature will be welcomed by users.

> I don't need this to be integrated. I don't have any intentions on wasting my time or your time. I have two working solutions and patching after update is done in minutes. If this is unwanted or unappreciated, I'm happy with closing and deletion. This is your call to make. After your suggestions two weeks ago I took time, to rework the code and test on current version 25 and version 26 beta 3 and 4. […] I took time, to rewrite the code and even more time to test it and cleanup the coding. Now I have to go back and do it again, because your suggestions were too cautious phrased. I'm not very happy with this and I will have to make time for this next weekend. 

I am happy with the refactoring you did regarding fetching the user and I do not think it overlaps that much with going through the account intead of `OCP\IUser`.
Only the calls to `$user->setProfileProperty` will need to change on user_ldap side.

> 2. What do you want me to do with those unset. I don't want you to be uncomfortable. I did, what I thought was intended. Looking at the original code of processAttributes, for example version 25 starting line 162, you can see a number of unset($attr) in lines 172, 192, 201, 232. I used that present code as guideline for my coding. I am pretty unaffected by wether or not those are present. I’m still not comfortable with those unset, we are not micromanaging memory anywhere else in our code like this. Like I said, You are doing it in your code exactly in the questioned processAttributes function. What should I do to make you comfortable? 3. I will take time, to ckeck and test this specifically and add more clearification to the documentation.

Thank you for the pointers on processAttribute, that’s code from before I joined Nextcloud, I did not see much unset in the rest of the code.
I was just afraid by merging those unset of setting a precedent and having unsets added everywhere in the code. Because I do not understand why these local vars should be unset more than any other.
@blizzz Do you remember what is the rationale behind the `unset` calls in `processAttribute`?

Thank you for your details on scope handling, that sounds quite reasonable.

If this is eating too much of your time I can take it from here if your prefer.

march42 committed:
refined code, to be independend from OCP\IUser

rework updateProfile in user_ldap/lib/User/User.php
some cleanup at processAttributes in user_ldap/lib/User/User.php
rearranged Fediverse attribute, to match profile layout

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
undoing the additions to OC\IUser

removed changes from lib/public/IUser.php
removed changes from lib/private/User/LazyUser.php
removed changes from lib/private/User/User.php

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 commented:
> I’m sorry you feel that way, that was not my intention of course.
> I’m used to reviewing code from collegues and being picky about it, to try to merge the best code as possible.
> Sorry if that felt like too much complains,

Thanks. I needed to address it anyways.
I'm here with you. I, myself too, love being picky about coding and following rules.
I felt like being pushed to decide things, not my decision to make. Espacially the `unset` commands. I would have just removed all of them, if you had just told me to do so. I was just not willing to make this decision on my own.

> I am happy with the refactoring you did regarding fetching the user and I do not think it overlaps that much with going through the account intead of `OCP\IUser`. Only the calls to `$user->setProfileProperty` will need to change on user_ldap side.

Thanks. Now I have done the rest, like you suggested and I removed the added code from IUser interface.

> If this is eating too much of your time I can take it from here if your prefer.

Thank you for the offer. I think I have it almost finished for now. I finished the code and finally did the rebase. For now, I'm testing and trying to find scenarios where it fails or crashes.

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +574,51 @@ private function verifyQuotaValue(string $quotaValue) {
+			return;
+		}
+		// prepare AccountManager and Account
+		$accountManager = \OC::$server->get(IAccountManager::class);
```
Is this a good way to get an accountManager instance, to work with?

come-nc replied:
```suggestion
		$accountManager = Server::get(IAccountManager::class);
```
This with a `use OCP\Server` at the beginning of the file is the preferred way now.
The best is to use DI (Dependency Injection) in the constructor, but here for a user object it’s more complicated because it’s not a singleton. In this instance I think it’s fine to use a get here.

march42 commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +574,51 @@ private function verifyQuotaValue(string $quotaValue) {
+			try {
+				$accountProperty = $account->getProperty($property);
+				$currentValue = $accountProperty->getValue();
+				$scope = ($profileScope ? $profileScope : ($accountProperty->getScope() ? $accountProperty->getScope() : AccountManager::DEFAULT_SCOPES[$property]));
```
I got the idea, that `AccountManager::DEFAULT_SCOPES[$property]` doesn't get changed by setting `account_manager.default_property_scope` in `config.php` - do I need to check this config setting extra?

come-nc replied:
DEFAULT_SCOPES is a const, it never changes.
You can see in AccountManager.php line 674 that it merges that with values from the config file to build the actual default scope.

Ideally you should not have to use AccountManager since it’s in `OC` and not `OCP`.
But it seems you found a missing feature from OCP here.
So either we’d need to add a getPropertyDefaultScope method to `OCP\IAccountManager`, or a magic value to give setProperty when wanting to use the default scope. (either null or a new IAccountManager::VISIBILITY_DEFAULT)
@Pytal What do you think, did I miss something?

march42 replied:
Looking into the code again, I think about removing the scope handling from my code and only use `Account::getProfileScope`

- I added this, to make sure all profile data stays on my server setting `v2-local`
- using the `config.php` `account_manager.default_property_scope` is more flexible and the user stays in control
- I want to rely on the defaults - using config.php and existing mechanisms
- new user will have empty profile with default scope applied
- if user changes property scope it will be used (CAN be both, pro and con ... and IS users choice)

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -512,6 +574,51 @@ private function verifyQuotaValue(string $quotaValue) {
+				$currentValue = $accountProperty->getValue();
+				$scope = ($profileScope ? $profileScope : ($accountProperty->getScope() ? $accountProperty->getScope() : AccountManager::DEFAULT_SCOPES[$property]));
+			}
+			catch (PropertyDoesNotExistException $e) { // thrown at getProperty
```
It can be thrown by getProperty if the property do not exist. You should leave it there, if Account is ever modified to remove one of the property you use we will get a clean handling instead of a misterious crash. Better safe than  sorry

come-nc reviewed (approved):
Apart from the tiny default scope question, all good!

Just tested and confirm it works.

march42 committed:
nice up the code handling AccountManager

merging defaultScopes from DEFAULT_SCOPES and account_manager.default_property_scope
removing unneccessary profileScope setting (using config.php instead)
honoring admin choice 'profile.enabled'=>false in config.php
moved checking for empty array to updateProfile function
corrected some typos and cleaned some comments

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Marc Hefter <marchefter@gmail.com>

Pytal commented:
Are the FIXMEs still to be done in this PR @march42?

march42 commented:
> Are the FIXMEs still to be done in this PR @march42?

@Pytal yes. I couldn't find a scenario, where those two checks against null in User.php (line 590 and 596) updateProfile will be true. I left them in there for "better safe than sorry" purpose.

Currently I'm doing some more testing and digging and found another question with profile handling. If I set phoneNumber to "0971..." in LDAP it gets formated to "+49971..." in user profile. And the next check in updateProfile (line 615) it get set again. I found this is not a big performance issue and think I need to add to documentation and clearify for the LDAP admin how the attributes should be prepared.
While checking with multi line attributes streetAddress and userBiography I noticed streetAddress has CRLF line ending and userBiography has a $ dollar sign instead of CR or LF. Here I need to double check again, where those difference comes from - OpenLDAP server or the configuration client. And I want to check Nextcloud, how multi line property should be formated here.
So, I was/am thinking about adding data formatting to processAttributes, at least for multi line values.

come-nc commented:
> Currently I'm doing some more testing and digging and found another question with profile handling. If I set phoneNumber to "0971..." in LDAP it gets formated to "+49971..." in user profile. And the next check in updateProfile (line 615) it get set again. I found this is not a big performance issue and think I need to add to documentation and clearify for the LDAP admin how the attributes should be prepared. While checking with multi line attributes streetAddress and userBiography I noticed streetAddress has CRLF line ending and userBiography has a $ dollar sign instead of CR or LF. Here I need to double check again, where those difference comes from - OpenLDAP server or the configuration client. And I want to check Nextcloud, how multi line property should be formated here. So, I was/am thinking about adding data formatting to processAttributes, at least for multi line values.

The $ thing is the postal address format in LDAP: https://www.rfc-editor.org/rfc/rfc4517#section-3.3.28
This is the complicated thing about allowing the admin to choose any attribute to map, there are many different syntaxes out there.

march42 commented:
> The $ thing is the postal address format in LDAP: https://www.rfc-editor.org/rfc/rfc4517#section-3.3.28
Yes, thank you, you're absolutely right. I did not remember this.

> This is the complicated thing about allowing the admin to choose any attribute to map, there are many different syntaxes out there.
And in most cases I would just tell the admin which syntax I want/need.
Last year I thought about implementing a formating syntax instead of plain attribute names. So the admin could format like "street+<CR>+postalCode+<SPACE>+l" and I decided this would be too much and dropped it.
My second thought was using the attributes syntax OID and mapping based on that. And right now I think this would be my preffered way to go.

BTW: now I can explain the reformatting of my phone numbers - E.123 format in LDAP allows many different syntaxes, i used international with dashes for separation, while E.164 in Nextcloud wants international without delimiters.

BUT: every checking and formatting here can/will have a performance impact.

march42 committed:
handling, formatting of postalAddress attribute syntax

replace '$' with ', ' delimiter for address property
reformatted some code to 80 column
early check and return, if wasRefreshed('profile')
removed FIXMEs after digging and double checking

Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 commented:
> Are the FIXMEs still to be done in this PR @march42?

@Pytal I finished and removed the FIXMEs.

march42 committed:
added simple data conversion for LDAP attributes

Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
trying to make github-code-scanning bot happy

Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 commented:
I changed my LDAP directory and have `nextcloud-user-profile.schema` a modified version of @blizzz `nextcloud.schema` (i think from the univention app) with attributes for all the fields added. I have no idea, where to put this or/and if I should ignore.

blizzz commented:
> I changed my LDAP directory and have `nextcloud-user-profile.schema` a modified version of @blizzz `nextcloud.schema` (i think from the univention app) with attributes for all the fields added. I have no idea, where to put this or/and if I should ignore.

That .schema file is an old format to define LDAP schemas, which may be imported to the LDAP server. This is nothing Nextcloud deals with.

march42 committed:
error handling in update profile from LDAP

added error message on InvalidArgumentException

Signed-off-by: Marc Hefter <marchefter@gmail.com>

march42 committed:
optimized handling of user profile data change

Check profile data checksum before updating user profile, to ensure
data has changed. Write checksum to user settings and cache.

Signed-off-by: Marc Hefter <marchefter@gmail.com>

szaimen commented:
Merge?

march42 commented:
> Merge?

If you're asking me, then yes.

come-nc commented:
There are open questions in https://github.com/nextcloud/documentation/pull/9614
